src/interface/swarm.hpp

```diff
@@ -47,6 +47,31 @@ struct BoundaryDeviceContext {
   ParticleBound *bounds[6];
 };
 
+// This class is returned by AddEmptyParticles. It provides accessors to the new particle
+// memory by wrapping the persistent new_indices_ array.
+class NewParticlesContext {
```

Comment:
This is equivalent to a view I guess, but it's cleaner, more clear design. So I like it.

Reply:
A view wouldn't give you the max index, though, because here in general the max index of new particles will be less than the preallocated new particle indices `ParArray`'s `size()`, right?

Anyway, the "contexts" pattern seems to be taking over for particles. It does seem pretty clear to me too.

Reply:
Ah ok yes I see what you mean. :+1: